GVRf/Framework/src/org/gearvrf/GVRHybridObject.java

```diff
@@ -185,6 +196,189 @@ public int getUseCount() {
+         *            List of cleanup handler(s)
+         */
+        protected CleanupHandlerListManager(
+                List<NativeCleanupHandler> prefixList) {
+            mPrefixList = prefixList;
+        }
+
+        /**
+         * Descendants that add a cleanup handler list use this method to create
+         * unique concatenations of their list with any of <em>their</em>
+         * descendants' list(s).
+         * 
+         * @param suffix
+         *            Descendant's (static) list
+         * @return A unique concatenation
+         */
+        protected List<NativeCleanupHandler> getUniqueConcatenation(
+                List<NativeCleanupHandler> suffix) {
+            if (suffix == null) {
+                return mPrefixList;
+            }
+
+            List<NativeCleanupHandler> concatenation = mUniqueCopies
+                    .get(suffix);
+            if (concatenation == null) {
+                concatenation = new ArrayList<NativeCleanupHandler>(
+                        mPrefixList.size() + suffix.size());
+                concatenation.addAll(mPrefixList);
+                concatenation.addAll(suffix);
+                mUniqueCopies.put(suffix, concatenation);
+            }
+            return concatenation;
+        }
+    }
+
+    private static class GVRReference extends PhantomReference<GVRHybridObject> {
+
+        // private static final String TAG = Log.tag(GVRReference.class);
+
+        private long mNativePointer;
+        private final List<NativeCleanupHandler> mCleanupHandlers;
+
+        private GVRReference(GVRHybridObject object, long nativePointer,
+                List<NativeCleanupHandler> cleanupHandlers) {
+            super(object, sReferenceQueue);
+
+            mNativePointer = nativePointer;
+            mCleanupHandlers = cleanupHandlers;
+        }
+
+        private void close() {
+            if (mNativePointer != 0) {
+                if (mCleanupHandlers != null) {
+                    for (NativeCleanupHandler handler : mCleanupHandlers) {
+                        handler.nativeCleanup(mNativePointer);
+                    }
+                }
+                NativeHybridObject.delete(mNativePointer);
+            }
+
+            sReferenceSet.remove(this);
+        }
+    }
+
+    private static class GVRFinalizeThread extends Thread {
+
+        // private static final String TAG = Log.tag(GVRFinalizeThread.class);
+
+        private GVRFinalizeThread() {
+            setName("GVRF Finalize Thread");
+            setPriority(MAX_PRIORITY);
+            start();
+        }
+
+        @Override
+        public void run() {
+            try {
+                while (true) {
+                    GVRReference reference = (GVRReference) sReferenceQueue
+                            .remove();
+                    reference.close();
+                }
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            }
+        }
+    }
+
```

Comment:
> Imho, I think the try catch block should be inside the while loop. 

But then we're entering and exiting a try/catch block every time GC adds a `PhantomReference` to the queue. This way, we only enter the try/catch block once.

> Otherwise, if any exception happens, the app will be still running while this GC thread has already stopped working.

But ... how is that going to happen? When we create the (self-starting) thread on line 216, we don't save any references to it. How are we going to interrupt a thread we have no reference to?

(We don't have to worry about the thread getting garbage collected, though: threads count as GC roots, just like static fields do.)